pkg/asset/agent/gencrypto/authconfig.go

```diff
@@ -128,3 +181,160 @@ func localJWTForKey(id string, privateKkeyPem string) (string, error) {
 
 	return tokenString, nil
 }
+
+func (a *AuthConfig) createAuthTokenSecret(kubeconfigPath string) error {
+	client, err := initK8sClient(kubeconfigPath)
+	if err != nil {
+		return err
+	}
+
+	authToken, err := a.getOrCreateAuthTokenFromSecret(client)
+	if err != nil {
+		return err
+	}
+	if authToken != "" {
+		// Update the token in asset store with the retrieved token from the cluster
+		a.AgentAuthToken = authToken
+	}
+	return nil
+}
+
+func initK8sClient(kubeconfig string) (*kubernetes.Clientset, error) {
+	var err error
+	var config *rest.Config
+	if kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	k8sclientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return k8sclientset, err
+}
+
+func (a *AuthConfig) getOrCreateAuthTokenFromSecret(k8sclientset kubernetes.Interface) (string, error) {
+	// check if secret exists
+	retrievedSecret, err := k8sclientset.CoreV1().Secrets(authTokenSecretNamespace).Get(context.Background(), authTokenSecretName, metav1.GetOptions{})
+
+	// if the secret does not exist
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Create the secret with the new JWT token having an exp claim of 48 hours
+			// i.e. the token will be valid for next 48 hours
+			err = createSecret(k8sclientset, a.AgentAuthToken)
+			if err != nil {
+				return "", err
+			}
+			return "", nil
+		}
+		// Other errors while trying to get the secret
+		return "", fmt.Errorf("failed to get required auth token secret from the cluster: %w. Re-run 'add-nodes' command to create new image files(ISO/PXE files)", err)
+	}
+
+	// the secret does exist
+	var authToken string
+
+	updatedAtStr := retrievedSecret.Annotations["updatedAt"]
+	updatedAt, err := time.Parse(time.RFC3339, updatedAtStr)
+	if err != nil {
+		return "", err
+	}
+	// if the secret with JWT token is older than 24 hours
+	// update the secret with a new JWT token with an exp claim of 48 hours
+	// i.e. the token will be valid for next 48 hours
+	if time.Since(updatedAt) > 24*expirationUnit {
+		err = updateSecret(k8sclientset, retrievedSecret, a.AgentAuthToken)
+		if err != nil {
+			return "", err
+		}
+		logrus.Debug("auth token secret regenerated and updated in the cluster")
+	} else {
+		logrus.Debug("auth token secret is still valid")
+		authToken, err = getSecret(retrievedSecret)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return authToken, err
+}
+
+func createSecret(k8sclientset kubernetes.Interface, newAgentAuthToken string) error {
+	createdAt := time.Now().UTC()
+	expiresAt := createdAt.Add(time.Duration(expiration) * expirationUnit)
+	// Create a Secret
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: authTokenSecretName,
+			Annotations: map[string]string{
```

Comment:
Do we need these annotations when the token itself has the expiration time?

Reply:
no, I will be pushing new changes